src/node/node_class.hpp

```diff
@@ -1093,118 +1082,80 @@ inline std::vector<Napi::PropertyDescriptor> ObjectWrap<ClassType>::create_napi_
 			node::String* name = get_cached_property_name(propName);
 			auto descriptor = Napi::PropertyDescriptor::Accessor<property_getter, property_setter>(name->ToString(env), napi_enumerable, (void*)name);
 			properties.push_back(descriptor);
+			// XXX should computed properties be reported as enumerable own keys?
 		}
 	}
 
 	return properties;
 }
 
 template<typename ClassType>
-Napi::Object ObjectWrap<ClassType>::create_instance_by_schema(Napi::Env env, Napi::Function& constructor, const realm::ObjectSchema& schema, Internal* internal) {
-	Napi::EscapableHandleScope scope(env);
+const Napi::FunctionReference& ObjectWrap<ClassType>::get_or_create_realm_object_factory(Napi::Env env, Napi::Function& constructor, const realm::ObjectSchema& schema, const SharedRealm& realm) {
+	static_assert(std::is_same_v<ClassType, realm::js::RealmObjectClass<realm::node::Types>>);
+
 	auto& s_schemaObjectTypes = get_schemaObjectTypes();
 	auto& s_class = get_class();
 
-	bool isRealmObjectClass = std::is_same<ClassType, realm::js::RealmObjectClass<realm::node::Types>>::value;
-	if (!isRealmObjectClass) {
-		throw Napi::Error::New(env, "Creating instances by schema is supported for RealmObjectClass only");
-	}
-
-	if (isRealmObjectClass && !internal) {
-		throw Napi::Error::New(env, "RealmObjectClass requires an internal realm object when creating instances by schema");
-	}
+	std::string path = realm->config().path;
+	std::string schemaName = schema.name + ":" + std::to_string(realm->schema_version());
 
-	Napi::Object instance;
-	auto config = internal->realm()->config();
-	std::string path = config.path;
-	auto version = internal->realm()->schema_version();
-	std::string schemaName = schema.name + ":" + std::to_string(version); 
+	auto& schemaObjects = s_schemaObjectTypes[path];
+	auto& schemaObject = schemaObjects[schemaName];
 
-	std::unordered_map<std::string, SchemaObjectType*>* schemaObjects = nullptr;
-	if (!s_schemaObjectTypes.count(path)) {
-		//std::map<std::string, std::map<std::string, SchemaObjectType*>>
-		schemaObjects = new std::unordered_map<std::string, SchemaObjectType*>();
-		s_schemaObjectTypes.emplace(path, schemaObjects);
-	}
-	else {
-		schemaObjects = s_schemaObjectTypes.at(path);
-	}
 
+	auto ownKeys = Napi::Array::New(env);
 	Napi::Function schemaObjectConstructor;
-	Napi::Symbol externalSymbol = ExternalSymbol;
+
 	//if we are creating a RealmObject from schema with no user defined constructor
 	if (constructor.IsEmpty()) {
-		//1.Check by name if the constructor is already created for this RealmObject 
-		if (!schemaObjects->count(schemaName)) {
-
-			//2.Create the constructor
-
-			//get or create the RealmObjectClass<T> constructor
+		//1.Check by name if the factory is already created for this RealmObject 
+		// XXX should we also check if `constructor` is empty?
```

Comment:
This seems like a bug in the old code. I kept the behavior for now, but I'm not sure I should have. In the case where the user passes a ctor in, we check that it is the *same* ctor, and rebuild the factory if not. Should we do the same if they switch from using a ctor to not using one?